File: rust/src/rpc.rs
use std::sync::Arc;

use tonic::{transport::Server, Request, Response, Status};
use tokio::sync::{mpsc, Mutex};
use tokio_stream::wrappers::ReceiverStream;
use triggered::Listener;


mod rpc_pb2 {
    tonic::include_proto!("rpc");
}
pub use rpc_pb2::*;

use crate::{
    runtime::Runtime,
    block::SignedBlock as SignedBlockInternal
};

type TransactionChannel = mpsc::Sender<Result<SignedTransaction, Status>>;
pub type TransactionQueues = Arc<Mutex<Vec<TransactionChannel>>>;
type BlockChannel = mpsc::Sender<Result<SignedBlock, Status>>;
pub type BlockQueues = Arc<Mutex<Vec<BlockChannel>>>;


pub struct RPC {
    port: String,
    runtime: Arc<Runtime>,
    transaction_queues: TransactionQueues,
    block_queues: BlockQueues,
    stop: Option<Listener>,
}

impl RPC {
    pub fn new(
        port: &u32,
        runtime: Arc<Runtime>,
        transaction_queues: TransactionQueues,
        block_queues: BlockQueues,
        stop: Listener,
    ) -> RPC {
        Self {
            port: port.to_string(),
            runtime,
            transaction_queues,
            block_queues,
            stop: Some(stop),
        }
    }

    pub async fn start(mut self) -> Result<(), Box<dyn std::error::Error>> {
        let addr = format!("[::1]:{}", self.port).parse()?;
    
        {
            let stop = self.stop.take().unwrap();
            Server::builder()
                .add_service(rpc_server::RpcServer::new(self))
                .serve_with_shutdown(addr, stop)
                .await?;
        }
    
        Ok(())
    }
}

#[tonic::async_trait]
impl rpc_server::Rpc for RPC {
    type TransactionFeedStream = ReceiverStream<Result<SignedTransaction, Status>>;
    type BlockFeedStream = ReceiverStream<Result<SignedBlock, Status>>;

    async fn transaction_feed(
        &self,
        _: Request<TransactionFeedRequest>,
    ) -> Result<Response<Self::TransactionFeedStream>, Status> {
        let (tx, rx) = mpsc::channel(1);
        self.transaction_queues.lock().await.push(tx);

        Ok(Response::new(ReceiverStream::new(rx)))
    }

    async fn block_feed(
        &self,
        _: Request<BlockFeedRequest>,
    ) -> Result<Response<Self::BlockFeedStream>, Status> {
        let (tx, rx) = mpsc::channel(1);
        self.block_queues.lock().await.push(tx);
        println!("PUSHED {:?}", self.block_queues.lock().await.len());

        Ok(Response::new(ReceiverStream::new(rx)))
    }

    async fn sync(
        &self,
        request: Request<SyncRequest>,
    ) -> Result<Response<BlockChain>, Status> {
        self.runtime.add_peer(&request.into_inner().address);        
        let reply = BlockChain {
            blocks: self.runtime.blockchain().lock().unwrap().blocks().iter().map(|b| b.pb()).collect(),
        };
        Ok(Response::new(reply))
    }
}

pub async fn get_blocks_from_peers(peers: &[String], address: &str) -> Result<Vec<SignedBlockInternal>, Box<dyn std::error::Error>> {
    let mut blocks = Vec::new();
    for peer in peers {
        blocks.clear();
        let mut client = rpc_client::RpcClient::connect(format!["http://{}", peer]).await?;
        let response = client.sync(Request::new(SyncRequest {address: address.to_owned()})).await?;
        blocks.extend(response.into_inner().blocks.iter().map(|b| SignedBlockInternal::from_pb(b)));
    }
    Ok(blocks)
}

File: python/src/proto_chain/rpc.py
from concurrent import futures
from queue import Queue, Empty

import grpc

from proto_chain import rpc_pb2, rpc_pb2_grpc


class RPC(rpc_pb2_grpc.RPCServicer):
    def __init__(self, runtime, transaction_queues, block_queues, stop_event):
        super().__init__()
        self._runtime = runtime
        self._transaction_queues = transaction_queues
        self._block_queues = block_queues
        self._stop_event = stop_event

    def TransactionFeed(self, request, context):
        queue = Queue()
        self._transaction_queues.append(queue)
        while not self._stop_event.is_set():
            try:
                yield queue.get(timeout=1).pb
            except Empty:
                ...

    def BlockFeed(self, request, context):
        queue = Queue()
        self._block_queues.append(queue)
        while not self._stop_event.is_set():
            try:
                yield queue.get(timeout=1).pb
            except Empty:
                ...

    def Sync(self, request, context):
        self._runtime.add_peer(request.address)
        return rpc_pb2.BlockChain(
            blocks=[b.pb for b in self._runtime.blockchain.blocks]
        )


def get_server(port, runtime, transaction_queue, block_queue, stop_event):
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    rpc_pb2_grpc.add_RPCServicer_to_server(RPC(runtime, transaction_queue, block_queue, stop_event), server)
    server.add_insecure_port(f'[::]:{port}')
    return server
